databricks/koalas/frame.py

```diff
@@ -6930,16 +6930,24 @@ def join(
             raise ValueError(
                 "columns overlap but no suffix specified: " "{rename}".format(rename=common)
             )
+
+        need_set_index = False
         if on:
+            if not is_list_like(on):
+                on = [on]  # type: ignore
+            if len(on) != right.index.nlevels:
+                raise ValueError(
+                    'len(left_on) must equal the number of levels in the index of "right"'
+                )
+
+            need_set_index = len(set(on) & set(self.index.names)) == 0
```

Comment:
@itholic Would you please help me understand this line?

Reply:
@xinrong-databricks 

Sure!

This line checks if the given join keys are already included in the `Index` or not.

If not (True, in this statement, because we're checking if the intersection count of the set is `0`), we need to set the given join keys as an `Index` using `set_index` below.

If you have any questions more, please feel free to ask ! :)

Reply:
Thank you! Your explanation is so clear :)!

May I ask for the reason for `set the given join keys as an Index`?

Reply:
This is because we uses `merge` after then.

If the given join keys are not in Index, the result of `merge` will be not correct.

For example, let's say we have two DataFrames as below.

```python
>>> kdf1
  key   A
0  K0  A0
1  K1  A1
2  K2  A2
3  K3  A3

>>> kdf2
      B
key
K0   B0
K1   B1
K2   B2
```

And we can expect the result of `join` with `on='keys'` as below.

```python
>>> kdf1.join(kdf2, on=['key'])
  key   A     B
0  K3  A3  None
1  K0  A0    B0
2  K1  A1    B1
3  K2  A2    B2
```

We can make this same result with `merge` as below.

```python
>>> kdf1.set_index('key').merge(kdf2, left_index=True, right_index=True, how='left').reset_index()
  key   A     B
0  K3  A3  None
1  K0  A0    B0
2  K1  A1    B1
3  K2  A2    B2
```

At this point, If we didn't `kdf1.set_index('key')`, the result will be different as below.

```python
>>> kdf1.merge(kdf2, left_index=True, right_index=True, how='left').reset_index()
   index key   A     B
0      0  K0  A0  None
1      1  K1  A1  None
2      3  K3  A3  None
3      2  K2  A2  None
```

So, that's why we need `set_index` here!

Reply:
Thank you @itholic ! That's so clear :).

Reply:
@xinrong-databricks np! Glad to know I helped :)